x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java

```diff
@@ -165,10 +207,30 @@ private SearchResponse buildResponse(long taskStartTimeNanos, InternalAggregatio
             skippedShards,
             tookInMillis,
             buildQueryFailures(),
-            clusters
+            clustersInfo.getClusters()
         );
     }
 
+    /**
+     * Useful for long running CCS async searches when using minimize-roundtrips
+     * @return true if the local cluster is being searched, all of the shards for the local search are completed
+     *              and the Clusters object is in a state where it should be updated to indicate that the local
+     *              cluster search has completed.
+     */
+    private boolean localClusterSearchIsFinished() {
+        if (clustersInfo.hasLocalShards() != Boolean.TRUE) {
+            return false;
+        }
+        int failedShards = buildQueryFailures().length;
+        Clusters clusters = clustersInfo.getClusters();
+        if (clusters == null  // should never be null for CCS minimizeRoundtrips scenario
+            // ensure that incrementing the 'successful' count will not cause successful+skipped to be larger than total
+            || (clusters.getSuccessful() + clusters.getSkipped()) >= clusters.getTotal()) {
+            return false;
+        }
+        return totalShards == (successfulShards + skippedShards + failedShards);
```

Comment:
I am afraid it is a bit delicate to look at number of shards to make conclusions around the local cluster being completed. See our docs (https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html#search-api-response-body ): 

> Note that shards that are not allocated will be considered neither successful nor failed. Having failed+successful less than total is thus an indication that some of the shards were not allocated.

Could we simplify things here by relying on the specific callback that gets called once the local cluster is done with its part? I guess the challenge is that `onFinalReduce` and `onPartialReduce` both call `updatePartialResponse`. I think that we could mark the local cluster done when onFinalReduce is called and we are minimizing roundtrips.